File: src/ui/main_window.rs
use crate::{Message, QuickAnalysisSource};
use iced::widget::{
    button, checkbox, column, container, progress_bar, row, scrollable, text, text_input, Column,
    Container, Row, Rule, Space,
};
use iced::{Alignment, Element, Length};

fn styled_container<'a>(content: impl Into<Element<'a, Message>>) -> Container<'a, Message> {
    container(content).padding(5)
}

fn group_box<'a>(title: &str, content: Column<'a, Message>) -> Column<'a, Message> {
    column![
        text(title).size(20),
        Rule::horizontal(1),
        container(content).padding(10)
    ]
    .spacing(5)
}

fn file_input_row<'a>(
    label: &'a str,
    value: &'a str,
    on_change: impl Fn(String) -> Message + 'a,
) -> Row<'a, Message> {
    row![
        text(label).width(Length::FillPortion(2)),
        text_input("Select file or directory...", value)
        .on_input(on_change)
        .width(Length::FillPortion(7)),
        button("Browse...").width(Length::FillPortion(1))
    ]
    .spacing(10)
    .align_items(Alignment::Center)
}

pub fn view<'a>(
    log_output: &'a str,
    status_text: &'a str,
    progress: f32,
    ref_path: &'a str,
    sec_path: &'a str,
    ter_path: &'a str,
    archive_logs: bool,
) -> Element<'a, Message> {
    let top_row = row![
        button("Settings...").on_press(Message::OpenOptions),
        Space::with_width(Length::Fill)
    ];

    let actions_group = group_box(
        "Main Workflow",
        column![
            button("Open Job Queue for Merging...")
            .on_press(Message::OpenJobQueue)
            .padding(10),
                                  checkbox("Archive logs to a zip file on batch completion", archive_logs)
                                  .on_toggle(Message::ArchiveLogsToggled)
        ]
        .spacing(10),
    );

    let analysis_group = group_box(
        "Quick Analysis (Analyze Only)",
                                   column![
                                       file_input_row("Source 1 (Reference):", ref_path, |text| {
                                           Message::QuickAnalysisInputChanged {
                                               source: QuickAnalysisSource::Reference,
                                               text,
                                           }
                                       }),
                                   file_input_row("Source 2:", sec_path, |text| {
                                       Message::QuickAnalysisInputChanged {
                                           source: QuickAnalysisSource::Secondary,
                                           text,
                                       }
                                   }),
                                   file_input_row("Source 3:", ter_path, |text| {
                                       Message::QuickAnalysisInputChanged {
                                           source: QuickAnalysisSource::Tertiary,
                                           text,
                                       }
                                   }),
                                   row![
                                       Space::with_width(Length::Fill),
                                   button("Analyze Only").on_press(Message::AnalyzeOnlyPressed)
                                   ]
                                   ]
                                   .spacing(10),
    );

    let status_bar = row![
        text("Status:"),
        Space::with_width(10),
        text(status_text).width(Length::Fill),
        progress_bar(0.0..=1.0, progress)
        .width(Length::Fixed(200.0))
    ]
    .align_items(Alignment::Center)
    .spacing(10);

    let results_group = group_box(
        "Latest Job Results",
        row![
            text("Source 2 Delay:"),
                                  text("—").width(100),
                                  text("Source 3 Delay:"),
                                  text("—").width(100),
                                  text("Source 4 Delay:"),
                                  text("—").width(100),
                                  Space::with_width(Length::Fill)
        ]
        .spacing(10),
    );

    let log_group = group_box(
        "Log",
        column![scrollable(text(log_output))],
    );

    let main_content = column![
        top_row,
        actions_group,
        analysis_group,
        status_bar,
        results_group,
        log_group,
    ]
    .spacing(15);

    styled_container(main_content).into()
}

File: crates/vsg_ui/src/cpp/main_window.cpp
#include "main_window.h"

#include <QGroupBox>
#include <QHBoxLayout>
#include <QVBoxLayout>

MainWindow::MainWindow(AppController* ctrl, QWidget* parent)
    : QMainWindow(parent), controller(ctrl) {
    setWindowTitle("Video/Audio Sync & Merge - Rust Edition");
    setGeometry(100, 100, 1000, 600);

    setupUi();
    connectSignals();

    // Set initial status
    statusLabel->setText("Ready");
}

void MainWindow::setupUi() {
    // Central widget
    auto* central = new QWidget(this);
    setCentralWidget(central);
    auto* mainLayout = new QVBoxLayout(central);

    // Top row with settings button
    auto* topRow = new QHBoxLayout();
    settingsBtn = new QPushButton("Settings...", this);
    topRow->addWidget(settingsBtn);
    topRow->addStretch();
    mainLayout->addLayout(topRow);

    // Main Workflow group
    auto* actionsGroup = new QGroupBox("Main Workflow", this);
    auto* actionsLayout = new QVBoxLayout(actionsGroup);

    jobQueueBtn = new QPushButton("Open Job Queue for Merging...", this);
    jobQueueBtn->setStyleSheet("font-size: 14px; padding: 5px;");
    actionsLayout->addWidget(jobQueueBtn);

    archiveLogsCheck = new QCheckBox("Archive logs to a zip file on batch completion", this);
    archiveLogsCheck->setChecked(true);
    actionsLayout->addWidget(archiveLogsCheck);

    mainLayout->addWidget(actionsGroup);

    // Quick Analysis group
    auto* analysisGroup = new QGroupBox("Quick Analysis (Analyze Only)", this);
    auto* analysisLayout = new QVBoxLayout(analysisGroup);

    // Source 1
    auto* source1Layout = new QHBoxLayout();
    source1Layout->addWidget(new QLabel("Source 1 (Reference):", this), 1);
    source1Input = new QLineEdit(this);
    source1Layout->addWidget(source1Input, 8);
    source1BrowseBtn = new QPushButton("Browse...", this);
    source1Layout->addWidget(source1BrowseBtn, 1);
    analysisLayout->addLayout(source1Layout);

    // Source 2
    auto* source2Layout = new QHBoxLayout();
    source2Layout->addWidget(new QLabel("Source 2:", this), 1);
    source2Input = new QLineEdit(this);
    source2Layout->addWidget(source2Input, 8);
    source2BrowseBtn = new QPushButton("Browse...", this);
    source2Layout->addWidget(source2BrowseBtn, 1);
    analysisLayout->addLayout(source2Layout);

    // Source 3
    auto* source3Layout = new QHBoxLayout();
    source3Layout->addWidget(new QLabel("Source 3:", this), 1);
    source3Input = new QLineEdit(this);
    source3Layout->addWidget(source3Input, 8);
    source3BrowseBtn = new QPushButton("Browse...", this);
    source3Layout->addWidget(source3BrowseBtn, 1);
    analysisLayout->addLayout(source3Layout);

    // Analyze button (right-aligned)
    auto* analyzeBtnLayout = new QHBoxLayout();
    analyzeBtnLayout->addStretch();
    analyzeBtn = new QPushButton("Analyze Only", this);
    analyzeBtnLayout->addWidget(analyzeBtn);
    analysisLayout->addLayout(analyzeBtnLayout);

    mainLayout->addWidget(analysisGroup);

    // Status row
    auto* statusLayout = new QHBoxLayout();
    statusLayout->addWidget(new QLabel("Status:", this));
    statusLabel = new QLabel("Ready", this);
    statusLayout->addWidget(statusLabel, 1);
    progressBar = new QProgressBar(this);
    progressBar->setRange(0, 100);
    progressBar->setValue(0);
    progressBar->setTextVisible(true);
    statusLayout->addWidget(progressBar);
    mainLayout->addLayout(statusLayout);

    // Latest Job Results group
    auto* resultsGroup = new QGroupBox("Latest Job Results", this);
    auto* resultsLayout = new QHBoxLayout(resultsGroup);
    resultsLayout->addWidget(new QLabel("Source 2 Delay:", this));
    source2DelayLabel = new QLabel("—", this);
    resultsLayout->addWidget(source2DelayLabel);
    resultsLayout->addSpacing(20);
    resultsLayout->addWidget(new QLabel("Source 3 Delay:", this));
    source3DelayLabel = new QLabel("—", this);
    resultsLayout->addWidget(source3DelayLabel);
    resultsLayout->addStretch();
    mainLayout->addWidget(resultsGroup);

    // Log group
    auto* logGroup = new QGroupBox("Log", this);
    auto* logLayout = new QVBoxLayout(logGroup);
    logOutput = new QTextEdit(this);
    logOutput->setReadOnly(true);
    logOutput->setFontFamily("monospace");
    logLayout->addWidget(logOutput);
    mainLayout->addWidget(logGroup);
}

void MainWindow::connectSignals() {
    // Connect browse buttons to controller
    connect(source1BrowseBtn, &QPushButton::clicked, [this]() {
        QString path = controller->browse_file(QString("Select Reference File"));
        if (!path.isEmpty()) {
            source1Input->setText(path);
        }
    });

    connect(source2BrowseBtn, &QPushButton::clicked, [this]() {
        QString path = controller->browse_file(QString("Select Secondary File"));
        if (!path.isEmpty()) {
            source2Input->setText(path);
        }
    });

    connect(source3BrowseBtn, &QPushButton::clicked, [this]() {
        QString path = controller->browse_file(QString("Select Tertiary File"));
        if (!path.isEmpty()) {
            source3Input->setText(path);
        }
    });

    // Connect analyze button
    connect(analyzeBtn, &QPushButton::clicked, [this]() {
        controller->analyze_sources(
            source1Input->text(),
            source2Input->text(),
            source3Input->text()
        );
    });

    // Connect stub buttons
    connect(settingsBtn, &QPushButton::clicked, [this]() {
        controller->open_settings();
    });

    connect(jobQueueBtn, &QPushButton::clicked, [this]() {
        controller->open_job_queue();
    });

    // Connect controller signals to UI slots
    connect(controller, &AppController::log_message, this, &MainWindow::onLogMessage);
    connect(controller, &AppController::progress_update, this, &MainWindow::onProgressUpdate);
    connect(controller, &AppController::status_update, this, &MainWindow::onStatusUpdate);
    connect(controller, &AppController::analysis_complete, this, &MainWindow::onAnalysisComplete);
}

void MainWindow::onLogMessage(const QString& message) {
    logOutput->append(message);
    // Auto-scroll to bottom
    logOutput->verticalScrollBar()->setValue(logOutput->verticalScrollBar()->maximum());
}

void MainWindow::onProgressUpdate(int percent) {
    progressBar->setValue(percent);
}

void MainWindow::onStatusUpdate(const QString& status) {
    statusLabel->setText(status);
}

void MainWindow::onAnalysisComplete(qint64 source2Delay, qint64 source3Delay) {
    source2DelayLabel->setText(QString("%1 ms").arg(source2Delay));
    source3DelayLabel->setText(QString("%1 ms").arg(source3Delay));
}
